libraries/node-core-library/src/FileSystem.ts

```diff
@@ -0,0 +1,432 @@
+   * @param linkTarget - The absolute or relative path to the target of the link.
+   */
+  public static createSymbolicLinkToFolder(linkSource: string, linkTarget: string): void {
+    // For directories, we use a Windows "junction".  On Unix, this produces a regular symlink.
+    fsx.symlinkSync(linkTarget, linkSource, 'junction');
+  }
+
+  /**
+   * Creates a symbolic link to a file (on Windows this requires elevated permissions).
+   * Behind the scenes it uses `fs.symlinkSync()`.
+   * @param linkSource - The absolute or relative path to the destination where the link should be created.
+   * @param linkTarget - The absolute or relative path to the target of the link.
+   */
+  public static createSymbolicLinkToFile(linkSource: string, linkTarget: string): void {
+    fsx.symlinkSync(linkSource, linkTarget, 'file');
+  }
+
+  /**
+   * Creates a hard link to a file.
+   * Behind the scenes it uses `fs.linkSync()`.
+   * @param linkSource - The absolute or relative path to the destination where the link should be created.
+   * @param linkTarget - The absolute or relative path to the target of the link.
+   */
+  public static createHardLinkToFile(linkSource: string, linkTarget: string): void {
+    fsx.linkSync(linkSource, linkTarget);
+  }
+
+  /**
+   * Follows a link to its destination and returns the absolute path to the final target of the link.
+   * Behind the scenes it uses `fs.realpathSync()`.
+   * @param linkPath - The path to the link.
+   */
+  public static followLink(linkPath: string): string {
+    return fsx.realpathSync(linkPath);
+  }
+
+  /**
+   * A helper function that converts line endings on a string.
+   * @param text - The text to be normalized.
+   * @param lineEndings - The style of line endings to use.
+   */
+  private static _convertLineEndings(text: string, lineEndings: NewlineConversion | undefined): string {
+    if (lineEndings === NewlineConversion.CrLf) {
+      return Text.convertToCrLf(text);
+    } else if (lineEndings === NewlineConversion.Lf) {
+      return Text.convertToLf(text);
+    }
+    return text;
+  }
+}
+
+/**
+ * API for interacting with file handles.
+ * @public
+ */
+export class File {
+  private _fileDescriptor: number | undefined;
+
+  /**
+   * Opens a new file handle to the file at the specified path and given mode.
+   * Behind the scenes it uses `fs.openSync()`.
+   * @param path - The absolute or relative path to the file handle that should be opened.
+   * @param flags - The flags for opening the handle
+   */
+  public static open(path: string, flags: FileHandleOpenFlags): File {
+    return new File(fsx.openSync(path, flags));
+  }
+
+  /**
+   * Writes some text to the given file handle. Throws if the file handle has been closed.
+   * Behind the scenes it uses `fs.writeSync()`.
+   * @param text - The text to write to the file.
+   */
+  public write(text: string): void {
```

Comment:
>write [](start = 9, length = 5)

If you were to add read(), would you use the same class?  If so what happens if you mix calls to read() and write()?

Or should the classes be split into FileReader and FileWriter?

Compare with C# StreamReader/StreamWriter

Reply:
We would split into 2.

---
In reply to: [198676267](https://github.com/Microsoft/web-build-tools/pull/727#discussion_r198676267) [](ancestors = 198676267)